### ISSUE TITLE
Improving WEBIRC

### ISSUE THREAD

**Issue by prawnsalad** [ircv3/ircv3-ideas#12]
WEBIRC is currently the only method to pass a users connection info to the IRCd when using some type of gateway service. For smaller setups this works well once setup, but for larger setups it starts to fall short.

Since running the Kiwi IRC webirc gateway for several years to support thousands of active networks and supporting hundreds of others to configure their own webirc gateways, these are the main issues that come up regularly:

* Configuration causes confusion and difficulties for a lot of admins.
* Whitelisting specific webirc gateways causes headaches as the gateway service scales or needs to change IP addresses.
* Networks still use identd lookups to a webirc gateway, which is pointless and slows down the connection from completing registration.
* Inconsistent results when using an incorrect webirc configuration.

It would be beneficial to discuss how we can improve this since the number of web client usage increases.

To kick start a discussion:

(For reference, the current WEBIRC implementation: `WEBIRC password gateway hostname ip`)

### Idea 1
* Use client cert auth to auth a webirc gateway instead of passwords. Most IRCds have cert auth code already so this should be trivial to implement. A good side effect of forcing TLS connections.
* Keep the current WEBIRC command structure but pass * for the gateway password since it is no longer needed.
* No longer whitelist specific webirc gateways
* Force a consistent ERROR message and disconnection on invalid webirc configurations

### Idea 2
* Replace the current static password in the WEBIRC command with a form of one time password.
* Use the gateway service name in the WEBIRC command to lookup any required secrets in the IRCd config.
* No longer whitelist specific webirc gateways
* Force a consistent ERROR message and disconnection on invalid webirc configurations

**Comment by ghost** [comment]
> No longer whitelist specific webirc gateways

How would this play out in terms of the specification itself? There's nothing in the `gateway <-> IRCD` negotiation that would prevent a network from deciding to lock gateways to specific IPs. While the specification could argue against whitelisting, it wouldn't break WebIRC itself to do so - the connection would either be allowed or denied in some form. As far as the latter is concerned, the gateway would never know it was whitelisted.

**Comment by prawnsalad** [comment]
Webirc gateway whitelisting isn't in the protocol itself, but currently every implementation (apart from maybe 1 IRCd that I know of) locks WEBIRC coming from specific IP ranges. Because of this is is recommended in the proposed spec to outline the current WEBIRC protocol.

What we could suggest though is that the whitelisting is not required, even SHOULD not be used. Otherwise implementations will continue the tradition of forcing a whitelist which will continue the pain point.

**Comment by dpyro** [comment]
To be clear, are you talking about an IRC equivalent of the `X-Forwarded-For` HTTP header?

**Comment by jobe1986** [comment]
Yes, WEBIRC is the IRC equivilent of HTTPs X-Forwarded-For where the IRC client uses the WEBIRC command to forward a clients real IP and host name to the IRCd.

The command itself takes the syntax "WEBIRC password username hostname ip" where current implimentations have the IRCd verifying the supplied password against their configuration. The 'username' parameter however is rarely if ever used for anything, and thus is mostly just ignored. And the 'hostname' parameter is usually either the uncloaked rDNS host name for the associated 'ip' or the same as 'ip' where an rDNS host name could not be determined.

As for where this idea is concerned, apart from standardising errors I personally don't see where a specification would even be appropriate for idea 1. As for idea 2, the specification would be more about a protocol by which to determine a one time password for use with the command, in which case such a protocol would be of use for other password commands such as PASS and OPER too so would be better as a generic spec providing a framework by which to determine a one time password.

**Comment by dpyro** [comment]
I'm not sure if this belongs here, but it would helpful to have a standardized rDNS host name exposed from the IRC server for WEBIRC clients. Each provider seems to have its own arbitrary format which can make it tricky for channel mods to create unique and localized bans.

**Comment by prawnsalad** [comment]
@dpyro There's a lot of politics involved in how some networks show the hostname for webirc users despite it being the main cause for ban evasion and confusion to users. I would be for suggesting that the hostname should be displayed as normal though.

**Comment by SaberUK** [comment]
@dpyro With IPv6 this problem is mainly solved as you can do what @irccloud do and give each user their own IP.

**Comment by kbuck1** [comment]
> it would helpful to have a standardized rDNS host name for WEBIRC clients

@dpyro, the solution is to ban WEBIRC clients by either username or realname (which well-behaved WEBIRC clients will set to the service name or web address) instead of the hostname. The WEBIRC protocol is designed to spoof the user's hostname to their real hostname, not some made-up one.

**Comment by DarthGandalf** [comment]
@SaberUK How is this related to IPv6? it's a different use, they can give each user an IP from one pool, which is not related to real user IP; WEBIRC exposes the real user IP.

**Comment by realJoshByrnes** [comment]
We currently use a non-standard version of the WEBIRC command, which is backwards compatible.

`WEBIRC password gateway hostname ip flags`
I'm not sure what flags would be useful, but we have a single flag - 's' (SSL/TLS)

Why? We have a reverse proxy which does the FCrDNS lookups, and translates incoming connections from various protocols (eg. TCP/TLS, HTTP/HTTPS, WS/WSS) to TCP.
Whilst we don't care which transport a user is using, it's important for us to know if they have a secure connection. We did consider implementing a second flag '6' for IPv6 connections, but decided against it. Thoughts?

I'm certainly all for **IDEA 1**, where a client certificate is used. Most current IRCd developers would be able to implement this with minimal changes, as most support TLS/SSL.

**Comment by sadiepowell** [comment]
>WEBIRC password gateway hostname ip flags

:+1: on the basic idea but I feel that using names with vendor prefixes (like caps and tags) would be better rather than using single letters to avoid the kind of collisions we get with modes.

**Comment by realJoshByrnes** [comment]
Something like this?
`WEBIRC password gateway hostname ip :flag1 flag2`

Eg. `WEBIRC * svc localhost 127.0.0.1 :secure irc7.com/vhost`
I think a secure flag should be vendor neutral, and MUST be implemented for the reasons stated above, where some clients may be connecting insecurely via a TLS connected webirc client.

**Comment by prawnsalad** [comment]
The flags are a good idea. And agreed with @SaberUK to use the same format as other parts of the protocol would be best. Using the caps+tags format would allow some room for project specific information.

A few standard tags that I think could be useful here:
* `secure` - as you mentioned this could be required if the user is using a secure connection
* `remote_port` - optional, may be handy in some cases
* `local_port` - optional, may be handy in some cases
* `geo` - optional, the gateway may do some GEOIP lookups

@ozjd Which gateway project is this that you're working on? I've been working on a very similar one that's being tested on a few networks at the moment, https://github.com/kiwiirc/webircgateway

**Comment by sadiepowell** [comment]
~~I just had a thought that this could be done with C2S tags rather than adding on new parameters...~~

~~`@webirc-secure=1 WEBIRC password gateway hostname ip`~~

**Comment by DanielOaks** [comment]
Hmm, I like adding the extra param which includes space-separated flags, it seems like it'd fall back alright for most current implementations. As much as I like C2S tags, sending them with `WEBIRC` to most servers that currently support it would prolly break them?

Need to do testing ofc and see how much the extra parameter would break current implementations of the command.

**Comment by prawnsalad** [comment]
WEBIRC is the first line to be sent so it must be backwards compatible which prevents adding the C2S tags.

**Comment by realJoshByrnes** [comment]
@prawnsalad O/T but the gateway is currently an internal one (closed source - although I will probably end up releasing it at some point). It's written in NodeJS and implements DNS lookups with the [FCrDNS](https://github.com/ozjd/FCrDNS) module that I've already published.

The C2S is not possible as it's unknown what the server can / will support.

**Comment by syzop** [comment]
**Idea 1** is a good one and we support this in UnrealIRCd (since about 2007). I'd definitely be in favor of it being promoted by WEBIRC gateways! I'm trying all I can to get more people to use TLS and like @prawnsalad said.. as a bonus this would help with that as well.

If you want to test you can use something like this (sha256 fingerprint of certificate) on UnrealIRCd 4.0.x:
```
webirc {
	mask *;
	password "B4:33:5F:68:D5:DC:E7:A8:6F:EC:74:66:C7:99:10:98:D7:DC:E1:B5:AE:5F:96:ED:4C:A1:EF:F3:B5:B7:35:28" { sslclientcertfp; };
};
```

UPDATE: I just realized that because we use I:line style matching you can only have one such webirc block. I'll fix that problem in next release (Oct 2017) so people who need multiple webirc { } blocks with sslclientcert/sslclientcertfp can use the feature as well.

**Comment by syzop** [comment]
The flags idea from @ozjd got me thinking...
If a WEBIRC gateway indicates 's' on a non-TLS connection then I wouldn't want to set a user as "Secure". After all, traffic is sent unencrypted to the WEBIRC gateway.

But let's turn it around. How about a WEBIRC gateway that uses a TLS connection from WEBIRC<->ircserv but uses an insecure connection such as HTTP to the end-user. Right now at UnrealIRCd, and I would assume at other IRCd's as well, we would call such a user "Secure", as we have no way to detect this situation.
So, two possible ways to "solve" this:
* Add some 'i' (insecure) or 'p' (plaintext) flag, so WEBIRC servers can indicate such a situation and call the connection "insecure" (non-TLS); OR
* Add the 's' (secure) flag as discussed earlier. For all servers that do not send this, even if they use an SSL/TLS connection, the users will show up as "insecure" (non-TLS).

The best way would be the latter option. I could unilaterally impose such a restriction from the UnrealIRCd side but I would really prefer some backing from WEBIRC and other IRCd software writers instead. After all, without support of the community, WEBIRC software could just always start sending 's' regardless of HTTP/HTTPS status and then all the hassle is for nothing.

Thoughts?

[Note that I'm only talking about how users show up on IRC.. so with umode +z, and in /WHOIS]

**Comment by syzop** [comment]
@jwheare: your thoughts on the above?

**Comment by prawnsalad** [comment]
> Add the 's' (secure) flag as discussed earlier. For all servers that do not send this, even if they use an SSL/TLS connection, the users will show up as "insecure" (non-TLS).

This was the idea. Although using optionally prefixed tags rather than individual letters so it's more inline with other specs. As for trusting gateways to be setting this secure flag correctly, well you already trust the gateway to be setting the correct hosts without malice so you should be able to trust that they use the secure flag correctly too. If not, then you remove the gateway from your network.

I can easily put this into the webircgateway project that networks are starting to make use of including kiwiirc since its backwards compatible, but it would need testing against different IRCDs to be sure of that first.

**Comment by jwheare** [comment]
@syzop I don't have a lot of useful thoughts about this as I've not really done much WEBIRC implementation. I'd rather delegate to others' experience at this stage. Good to see further discussion of these issues though.

**Comment by realJoshByrnes** [comment]
@syzop Our current implementation will only set flag 's' if the user is connecting to the WebIRC compatible proxy via HTTPS/TLS. In this case, the WebIRC <-> IRCd connection is sent over TLS. 

* If the 's' flag is not received by the IRCd, the client MUST be considered insecure.
* If the 's' flag is received by the IRCd over an insecure port, the client MUST be considered insecure.

It should be noted that the flag 's' does not contain transport specific information, and could be SSL/TLS/something else in the future.

As for WebIRC sending 's' for insecure clients - that is a possibility, but we already have a high level of trust required for WebIRC servers.

**Comment by syzop** [comment]
> I can easily put this into the webircgateway project that networks are starting to make use of including kiwiirc

Great!

And what you @prawnsalad, @ozjd and @DanielOaks all say:
> I like adding the extra param which includes space-separated flags, it seems like it'd fall back alright for most current implementations.

There's a clear consensus towards tokens here. I would also prefer that because it's easy to expand. So let's go for that.

Regarding breaking servers (which applies to any modification, flags or tokens). I guess someone would have to test this with a couple of server implementations. Testing is the only way to tell...
I just checked UnrealIRCd and it has no problem with the extra parameters. If I look at the source from 10 years ago, it also has no problems with it. Key here is it adds the command as *MAXPARA* and not with *4*.

> It should be noted that the flag 's' does not contain transport specific information, and could be SSL/TLS/something else in the future.

Right. How about, if we go for tokens, we call it 'tls'? That has no ambiguity.

**Comment by syzop** [comment]
And sorry @ozjd for reading over your comment earlier:
> I think a secure flag should be vendor neutral, and MUST be implemented for the reasons stated above, where some clients may be connecting insecurely via a TLS connected webirc client.

I now indeed see you already proposed exactly what I wrote a few comments up. I didn't mean to (appear to) claim your idea :D

**Comment by nomis** [comment]
If a WEBIRC gateway accepts users over HTTP then it shouldn't be allowed to authenticate to the network at all, so I think these flags are unnecessary.

**Comment by syzop** [comment]
If you mean that WEBIRC gateways should use HTTPS then I agree. Just like all IRC connections should use TLS. Unfortunately the world is less perfect...
The thing we are trying to fix here is the misleading case where a WEBIRC gateway accepts HTTP users but connects through TLS to the IRC server. I doubt you can fix that by simply saying "don't do that". I think adding the proposed "tls" flag to the "WEBIRC" command would be a better and safer solution.

**Comment by sadiepowell** [comment]
>If a WEBIRC gateway accepts users over HTTP then it shouldn't be allowed to authenticate to the network at all, so I think these flags are unnecessary.

What *should* happen and what *does* happen are two different things.

**Comment by nomis** [comment]
So you're running networks that allow any WEBIRC gateway to connect? No.

If you're running a network that cares enough about security to have TLS connections to the IRC server then your WEBIRC gateway website will support HTTPS.

**Comment by realJoshByrnes** [comment]
IMHO the IRCd should not control the protocols that are accepted by WebIRC. As our WebIRCish service is a reverse proxy, all incoming connections are handled - including plain TCP. 

@nomis : "gateway website will support HTTPS." - supporting HTTPS doesn't mean declining HTTP.

**Comment by syzop** [comment]
> IMHO the IRCd should not control the protocols that are accepted by WebIRC. As our WebIRCish service is a reverse proxy, all incoming connections are handled - including plain TCP.

I don't understand. I thought you were in favor?

**Comment by realJoshByrnes** [comment]
I am in favour of the flags, but I  don't agree with @nomis suggestion of limiting the WebIRC service's protocols.

**Comment by syzop** [comment]
Ahh okay. I was reading it in a different context. For example at UnrealIRCd we have a plaintext-policy setting which you could set to reject plaintext clients. In such a case it should also reject connections without HTTP (detected via lack of tls flag in WEBIRC). It's currently rarely done but who knows later on...

But you were strictly talking about @nomis suggestion of disallowing all WEBIRC gateways to use HTTP (or anything else without TLS). Yeah. I agree. Certainly for reverse proxies and load balancers.

To get back at my plaintext-policy example... I think if you don't trust plaintext clients then you shouldn't only disallow/reject HTTP, but all of them including directly connected plaintext clients (with no WEBIRC involved). Only disallowing WEBIRC's that use HTTP makes only marginal sense IMO.

Anyway, I'm/we are kinda repeating ourselves here. Agree to disagree @nomis?

**Comment by sadiepowell** [comment]
InspIRCd master now has support for authenticating WebIRC connections with client certificate fingerprints either with or without a password depending on configuration.

I have started implementing @prawnsalad's attribute proposal over here too: https://github.com/inspircd/inspircd/compare/master...SaberUK:master+webirc-attribs

**Comment by DanielOaks** [comment]
@SaberUK I'm implementing the same [over here](https://github.com/oragono/oragono/commit/b0649cb5d354f6ef5772c519b688b55aa3b85586), can we get the flags proposal thrown into the main repo as a PR on the WebIRC spec?

**Comment by syzop** [comment]
@prawnsalad & others: has it been tested against a couple of (major) IRCd's to see if this is backwards compatible? That would be the only (major) showstopper.
It's easy to make a fixed paracount vs MAXPARA "mistake" on the IRCd-side, hence why I ask.

**Comment by prawnsalad** [comment]
@syzop As a quick initial test I've run it against the following without issue:
* ircu
* inspircd
* Unreal3.2.6
* charybdis-3.5.0

I've now started adding the extra tags into webircgateway here, https://github.com/kiwiirc/webircgateway/commit/dfc0f74c5769743f8c9fccb32e577b9c01b86445

This can be tested with the new kiwi client here https://kiwiirc.com/nextclient/

One thing that has come up while tlaking with @SaberUK is the formatting of the tags. We can either use the exact same formatting as the existing tags, ie. separated by `;`, or keep them space separated and throw them all into the trailing parameter. Personally I'm for keeping things consistent and matching the existing tags exactly - separated by `;`.

I'm going to draft up a spec this weekend so any test implementations can be working towards the same format + standard tag names.

**Comment by syzop** [comment]
Thanks @prawnsalad for testing a couple of IRCd's. That was a potential show-stopper.

As for syntax:
I'd go with space separated like @DanielOaks, @ozjd and InspIRCd suggested. It's how most things are done at IRC, including in CAP (another handshake message), ISUPPORT, ISON and many many others.
The ```;``` as a separator I've never seen before in the IRC protocol, it seems to only exist in *message-tags*.
The only other common alternative is ```,```. However, with space you can still use comma in any parameters (property=x,a) so space separated is much more preferable.

So, to quote ozjd's comment from July. IMO we should go with this syntax:
```WEBIRC password gateway hostname ip :flag1 flag2```

**Comment by jobe1986** [comment]
I can also confirm that the additional parameters are silently ignored on Nefarious without issue and allowing the base WEBIRC command to still work as expected.

**Comment by prawnsalad** [comment]
Before typing up the draft spec for this, the discussion on the tags format should be discussed properly so that it's in-line with other parts of the protocol - existing and new. https://github.com/ircv3/ircv3-ideas/issues/19

**Comment by syzop** [comment]
AFAICT everyone wants space separated here but you.

**Comment by prawnsalad** [comment]
@syzop feel free to mention that on #19. Hoping to get more discussion there to sync up all the different kay/val structures we currently have

**Comment by jwheare** [comment]
I added examples of various k/v formats in #19.

Does WEBIRC need to support values with spaces or complex nested types in flags? If so, I'd suggest going with message tag format. If not, might be better to go with something simpler like CAP or ISUPPORT.

**Comment by prawnsalad** [comment]
Spaces not needed right now, but it's more 1) keeping it available in case spaces are needed in future (we don't want to lock ourselves out of future options), and 2) being consistent with other parts of the protocol as mentioned in #19

**Comment by jwheare** [comment]
I don't know if full consistency is always achievable. The escaping/delimiting rules for the various formats are specific to where they occur in the wire format. Since you're introducing extra flags in the last parameter space of the WEBIRC command, it would probably make most sense to follow the conventions of other messages that exist in that space, like CAP and ISUPPORT.

Imposing all of the escaping conventions from message tags doesn't really make sense if you aren't actually using message tags.

**Comment by prawnsalad** [comment]
While I do think it's a very bad idea - design and functionally - to lock out any future backwards compatible additions to a command (isn't that why we do all the hacks to existing IRC commands now?) and permanently blocking out characters for future use, it seems everyone else is perfectly fine with this and a PR spec already underway so lets go with that then.

**Comment by DanielOaks** [comment]
#19 pretty clearly needs to be resolved. It's been on my todo list for a while (and I'm guessing some other peoples' lists as well). However, it's gonna take a while to get that resolved because it just takes work, compiling the various escaping methods that we use around the place, making everything use it and etc (some also need to be clarified in ways, examples added, etc).

It's pretty clear that most of the people in here are in favour of a space-separated format for these flags, so that's what I went with for https://github.com/ircv3/ircv3-specifications/pull/328 . The thing about that PR is that there's no values that _need_ to be escaped in it, so no need for any escaping method right now.

We can continue churning and working on #19, in the meantime https://github.com/ircv3/ircv3-specifications/pull/328 can be worked on and merged, and then when we've got an appropriate escaping method we can put that into the WebIRC spec as well.

It's not a matter of locking out backwards-compatible additions, it's a matter of we don't need the escaping right now, and we can get to work implementing the feature while we work out the escaping part on the spec side.

**Comment by syzop** [comment]
UnrealIRCd 4.0.17 will no longer give user mode +z (=secure) to users on WEBIRC gateways using SSL/TLS IRC, unless the WEBIRC gateway gives us some assurance that the client<->webirc gateway connection is also secure (eg: https).
(4.0.17 is scheduled for +/- end of December)

**Comment by sadiepowell** [comment]
It would be nice if we could standardise key names before people start shipping stuff. Or, failing that, prefix them with a vendor prefix or `draft/`.

**Comment by syzop** [comment]
> It would be nice if we could standardise key names before people start shipping stuff. 

The thing is, this isn't an ordinary feature enhancement. It is about solving a security (loop)hole. Otherwise I could wait for many months...

On-topic:
@ozjd and @prawnsalad previously, and @DanielOaks's draft/PR, use "secure" and I'm fine with that.

**Comment by realJoshByrnes** [comment]
Just to be fair, we are currently using +s, but have no issues with +secure being implemented. We tried to stick with mode style +s.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by prawnsalad** [comment]
@DanielOaks, were you intending to re-open your PR at some point? It's currently put this discussion in limbo.

The names I'd put into the webircgateway as a live-test were `secure` and `remote-port`. As for the encoding we have 2 options.
1. Don't support it now simply because we don't have an immediate use. If 6 months down the line someone finds a use for a value with a space they are then fucked unless we either spend another few months drafting a new spec + waiting for IRCds to support it, or using yet another encoding for that specific value.
2. Add some future proofing into the spec we draft up now to support the use case without any fuss further down the line. This will be handy for people experimenting with what data is handy to be sent to the IRCd.
    1. Message tags format
    2. ISUPPORT token format

At this particular point I just want to get this sorted so I can I start making use of this in the webircgateway project in a standard way. Hopefully in such a way that this topic doesn't need to be revisited in a few months if I wanted to send a space character.

**Comment by DanielOaks** [comment]
My plan is to go with (2), using pretty much (or exactly) the same escaping we use for message tags (but having explicitly space-separated flags).

I'm planning to re-open the PR, it was closed because we couldn't come to a consensus on the escaping. If someone wants to write up the specific escaping section so I can slot it in or do something similar, I'll re-open it with those changes. Otherwise, feel free to take it, add a few commits on top and then submit a new PR based on it.

**Comment by RyanSquared** [comment]
Should this be closed now, as the PR seems to be merged?

**Comment by DanielOaks** [comment]
Yeah definitely, the PR's opened so we'll continue discussion in there

**Comment by jwheare** [comment]
Closed in favour of ircv3/ircv3-specifications#328
